### DISCUSSION TITLE
Morecantile.js a possible community need?

### DISCUSSION THREAD

**Discussion by AndrewAnnex** [developmentseed/morecantile#153] [General]
I have been doing some work on my [planetcantile](https://github.com/AndrewAnnex/planetcantile) project that provides a TiTiler endpoint with additional/not official (yet) TileMatrixSets for other planetary bodies like the Moon and Mars. As part of that effort I am currently updating [tilebench](https://github.com/AndrewAnnex/tilebench/tree/add_tms_option) and I am running into an issue where the viz tool requests the wrong tiles given the tilematrixset.

The reason for that seems to be the javascript within tilebench that does the lon/lat to tile id conversion. While I think I can solve that particular problem myself within that code, I realized that the community could benefit from a Javascript version of Morecantile that is able to perform the tile id and coordinate transformations to allow front ends like maplibre/leaflet/openlayers to use tilematrixsets natively in JS, and could simplify adding support for other TMS specs in these tools. 

I'd expect this new library to provide a subset of the capabilities of morecantile due to limitations in proj4js, so it could be a pretty light weight dependency in the end if scoped right. 

Thoughts?

**Comment by vincentsarago** [comment]
@AndrewAnnex I think you're talking about the usage of `sperical-mercator.js` within the tilebench template.

https://github.com/developmentseed/tilebench/blob/2dae78bc0e13cbcd0781258bb58998afa0ab5fd4/tilebench/templates/index.html#L170-L220

This could be easily replace by a new endpoint within the tilebench application (which will then use morecantile) instead of trying to do this (creating the geojson tile grid) in the frontend

I would recommend to do this (updating tilebench) in two steps: 1. switch to leaflet 2. add TMSgrid endpoint

**Reply by vincentsarago** [reply]
feel free to start and issue in tilebench 🙏

**Reply by AndrewAnnex** [reply]
@vincentsarago While I agree that is the fix for that particular issue in tilebench (both ideas sound fine for that tilebench pr), when looking at this issue this raised the idea for a "morecantile.js" library to perform the coordinate to tile index transformations in the browser. The specifics for tilebench weren't so much the motivation for this idea/question really.

For example functions like `tile`, `tiles`, `bounds`, `quadkey`, `quadkey_to_tile` and others could be useful to have in JS, and working with a similar API to morecantile would be attractive. 

I guess part of the issue against this is that for Earth built-in functionality in leaflet and other libraries is already sufficient, and anything more complicated would require changes upstream. My hope would be that this JS library could serve as that dependency for those packages to make it easier for them to use other tms specs.

**Reply by AndrewAnnex** [reply]
now maybe that just isn't something too many people need, but I wanted to raise the question to better understand the issue

**Reply by AndrewAnnex** [reply]
was also seeing https://github.com/missinglink/s2js recently as a source of why this could be useful to others
